compiler/rustc_lint_defs/src/builtin.rs

```diff
@@ -4175,3 +4176,27 @@ declare_lint! {
     Warn,
     "\"invalid_parameter\" isn't a valid argument for `#[macro_export]`",
 }
+
+declare_lint! {
+    /// The `unknown_diagnostic_attribute` lint detects unrecognized diagnostic attributes.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// #![feature(diagnostic_namespace)]
+    /// #[diagnostic::does_not_exist]
+    /// struct Foo;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// It is usually a mistake to specify a diagnostic attribute that does not exist. Check
+    /// the spelling, and check the diagnostic attribute listing for the correct name. Also
+    /// consider if you are using an old version of the compiler, and the attribute
+    /// is only available in a newer version.
+    pub UNKNOWN_DIAGNOSTIC_ATTRIBUTE,
```

Comment:
This should be plural as per [RFC 344](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints).

I also wonder what to do about diagnostic attributes that use a different and unknown format. E.g. previously one would only be allowed to specify `#[diagnostic::on_unimplemented = "abc def"]`, but now one can do `#[diagnostic::on_unimplemented(note = "hello", help = "some help note here")]`, or even `#[diagnostic::on_unimplemented(fluent_slug = "some_name")]`. What lint would then be fired, and if it is this lint, which is fine, is the name good?

Reply:
I've fixed the lint name in https://github.com/rust-lang/rust/pull/111780/commits/4f20a4f3f95355f692fa30c406d0d6914935b3b5

As for the unknown options variant: I would expect that we also implement a (different) lint there. At least as crate author I might want to allow new/changed options but no unknown attributes. This other lint can only be implemented after there is an actual attribute.

Reply:
@weiznich thanks! The proposed strategy to have two lints sounds good.